File: Minimum Window Substring.cpp
class Solution {
public:
    string minWindow(string S, string T) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        int ns = S.length();
        int nt = T.length();
        int num[256] = {0};
        int need[256] = {0};
        int count = 0;
        int minWin = INT_MAX, minBegin, minEnd;
        
        for (int i = 0; i < nt; i++) {
            num[T[i]]++;
        }
        
        for (int beg = 0, end = 0; end < ns; end++) {
            if (num[S[end]] == 0) continue;
            
            need[S[end]]++;
            if (need[S[end]] <= num[S[end]]) {
                count++;
            }
            
            if (count == nt) {
                while (num[S[beg]] == 0 || need[S[beg]] > num[S[beg]]) {
                    if (need[S[beg]] > num[S[beg]]) {
                        need[S[beg]]--;
                    }
                    beg++;
                    
                }
                
                if (minWin > end-beg+1) {
                    minWin = end-beg+1;
                    minBegin = beg;
                    minEnd = end;
                }
            }
        }
        
        if (count < nt) {
            return "";
        } else {
            return S.substr(minBegin, minWin);
        }
        
    }
};

File: Minimum Window Substring.cs
public class Solution {
    public string MinWindow(string s, string t) {
        if (string.IsNullOrEmpty(t) || string.IsNullOrEmpty(s)) {
            return "";
        }
        
        var counts = new int[256];
        foreach (var ch in t) {
            counts[ch]++;
        }
        
        var start = 0;
        var covers = new int[256];
        var cnt = 0;
        var minWin = s.Length;
        var minBeg = 0;
        for (var i = 0; i < s.Length; i++) {
            var ch = (int) s[i];
            if (counts[ch] == 0) {
                continue;
            }
            
            covers[ch]++;
            if (covers[ch] <= counts[ch]) {
                cnt++;
            }
            
            if (cnt == t.Length) {
                while (covers[s[start]] == 0 || covers[s[start]] > counts[s[start]]) {
                    if (covers[s[start]] != 0) {
                        covers[s[start]]--;
                    }
                    
                    start++;
                }
                
                if (minWin > i - start + 1) {
                    minWin = i - start + 1;
                    minBeg = start;
                }
            }
        }
        
        if (cnt < t.Length) {
            return "";
        } else {
            return s.Substring(minBeg, minWin);
        }
    }
}
